src/compiler/checker.ts

```diff
@@ -11182,7 +11180,11 @@ namespace ts {
                     return getSubstitutionType(maybeVariable as TypeVariable, instantiateType((<SubstitutionType>type).substitute, mapper));
                 }
                 else {
-                    return maybeVariable;
+                    const sub = instantiateType((<SubstitutionType>type).substitute, mapper);
+                    if (sub.flags & TypeFlags.AnyOrUnknown || isTypeSubtypeOf(getRestrictiveInstantiation(maybeVariable), getRestrictiveInstantiation(sub))) {

```

Comment:
When would the `isTypeSubtypeof` check be true?

Reply:
When `typeVariable = string[]` and `substitute = any[] & string[]`, for example. Likewise when `typeVariable = { a, b }` and `substitute = { a }`.

Reply:
Pretty much cases where supertype reduction in the substitute would be nice. :)

It's worth noting that doing this actually brings us back to pre-3.4 behavior for our output olwhen conditionals simplify. While the intersections with extra supertypes are generally equivalent, they delete information when they contain `any` which can poison further comparisons, so including them in the conditional results was altering some outputs. There was another issue where someone was mentioning it, but I think we marked it as intentional? I can't find the exact issue, but a user had a condition like `T extends I<any> ? T : never` - with `T` as `I<number`, we produced `I<number> & I<any>` - we'd actually broadened the type with the intersection rather than narrowed it. This supertype reduction check fixes that issue (and some of our test cases which changed to reflect the intersections when we added the simplifications).

Reply:
@weswigham This one I think #30489